internal/channel/channel.go

```diff
@@ -20,109 +18,155 @@ type Channel struct {
-	if err != nil {
-		return fmt.Errorf("failed to create stdin pipe: %w", err)
-	}
+// initPlugin returns a new Plugin or nil if an error occurred during initialization
+func (c *Channel) initPlugin() *Plugin {
+	c.Logger.Debug("Initializing channel plugin")
 
-	reader, err := cmd.StdoutPipe()
+	p, err := NewPlugin(c.Type, c.Logger)
 	if err != nil {
-		return fmt.Errorf("failed to create stdout pipe: %w", err)
+		c.Logger.Errorw("Failed to initialize channel plugin", zap.Error(err))
+		return nil
 	}
 
-	errPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stderr pipe: %w", err)
+	if err := p.SetConfig(c.Config); err != nil {
+		c.Logger.Errorw("Failed to set channel plugin config, terminating the plugin", zap.Error(err))
+		p.Stop()
+		return nil
 	}
 
-	go forwardLogs(errPipe, c.Logger)
+	return p
+}
 
-	if err = cmd.Start(); err != nil {
-		return fmt.Errorf("failed to start cmd: %w", err)
-	}
-	c.Logger.Debug("Cmd started successfully")
+// runPlugin is called as go routine to initialize and maintain the plugin by receiving signals on given chan(s)
+func (c *Channel) runPlugin() {
+	var currentlyRunningPlugin *Plugin
+	for {
+		if currentlyRunningPlugin == nil {
+			currentlyRunningPlugin = c.initPlugin()
+		}
 
-	c.cmd = cmd
-	c.rpc = rpc.NewRPC(writer, reader, c.Logger)
+		var rpcDone <-chan struct{}
+		if currentlyRunningPlugin != nil {
+			rpcDone = currentlyRunningPlugin.rpc.Done()
+		}
 
-	if err = c.SetConfig(c.Config); err != nil {
-		go c.terminate(c.cmd, c.rpc)
+		select {
+		case <-rpcDone:
+			c.Logger.Debug("Receive plugin error signal")
 
-		c.cmd = nil
-		c.rpc = nil
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-		return fmt.Errorf("failed to set config: %w", err)
-	}
-	c.Logger.Debugw("Successfully set config", zap.String("config", c.Config))
+			continue
+		case <-c.newConfigCh:
+			c.Logger.Debug("Receive reload plugin config signal")
 
-	return nil
-}
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-func (c *Channel) Stop() {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+			continue
+		case <-c.stopPluginCh:
+			c.Logger.Debug("Receive plugin stop signal")
 
-	if c.cmd == nil {
-		c.Logger.Debug("channel plugin has already been stopped")
-		return
-	}
+			if currentlyRunningPlugin != nil {
+				currentlyRunningPlugin.Stop()
+				currentlyRunningPlugin = nil
+			}
 
-	go c.terminate(c.cmd, c.rpc)
+			return
+		case req := <-c.notificationCh:
+			if currentlyRunningPlugin == nil {
+				currentlyRunningPlugin = c.initPlugin()
+			}
 
-	c.cmd = nil
-	c.rpc = nil
+			if currentlyRunningPlugin == nil {
+				c.Logger.Debug("Cannot send notification, plugin could not be started")
+				req.errCh <- errors.New("plugin could not be started")
+			} else {
+				go func(p *Plugin) {
+					req.errCh <- p.SendNotification(req.req)
+				}(currentlyRunningPlugin)
```

Comment:
Yes, the channel must be buffered.

Reply:
Currently, `p.SendNotification()` does all the RPC stuff and waits for the result. So it can basically take as long as a plugin might take (at the moment, there's also no timeout).